thewalrus/quantum.py

```diff
@@ -1139,3 +1143,50 @@ def update_probabilities_with_loss(etas, probs):
         qein = np.einsum(einstrings, loss_mat(eta, cutoff), probs)
         probs = np.copy(qein)
     return qein
+
+
+def update_probabilities_with_noise(probs_noise, probs):
+    """Given a list of noise probability distributions for each of the modes and a tensor of
+    probabilitites, calculate an updated tensor of probabilities after noise is applied.
+
+    Args:
+        probs_noise (list): List of probability distributions describing the noise in each of the modes
+        probs (array): Array of probabilitites in the different modes
+
+    Returns:
+        array: List of noise-updated probabilities with the same shape as probs.
+    """
+    probs_shape = probs.shape
+    num_modes = len(probs_shape)
+    cutoff = probs_shape[0]
+    if num_modes != len(probs_noise):
+        raise ValueError(
+            "The list of probability distributions probs_noise and the tensor of probabilities probs have incompatible dimensions."
+        )
+
+    @jit(nopython=True)
+    def update_1d(probs, one_d): # pragma: no cover

```

Comment:
How come this function is nested? Does it cause issues with the just in time compilation? (that is, since the inner function is redefined each time the outer function is called)

Reply:
It is now an independent (private) function